### ISSUE TITLE
Bug: Candid encoding not correctly assigning different variants for a `key: variant val` vector

### ISSUE THREAD

**Issue by ozwaldorf** [dfinity/candid#388]
for some reason dfx and didc is not working correctly for encoding a vec with the value which is a variant of different types

For simplicity, I stripped the argument down to only the offending structure, and provide an example with didc (didc 0.1.4) which is enough to replicate the issue

```
didc decode \
   "$(didc encode '(
      vec {
          record { "name"; variant { TextContent = "blah" } };
          record { "ttl"; variant { Nat64Content = 1 : nat64 } };
      }
   )')"
```
running this will give you an error:
```
Error: Trailing value after finishing deserialization

Caused by:
    input: 4449444c036d016c02007101026b01ecbfa8c90d71010002046e616d650004626c61680374746c000100_000000000000
    table: type table0 = vec table1
    type table1 = record { text; table2 }
    type table2 = variant { 3_643_416_556 : text }
    wire_type: text, expect_type: text
```

Which it seems to encode everything as that first type it comes across, instead of corresponding variants, as seen by table2

### Meta

`dfx --version`:
```
dfx 0.11.2
```

`didc --version`:
```
didc 0.1.4
```

**Comment by sesi200** [comment]
Thank you for the report with such a nice repro! Dfx contains no code of its own for this operation - it uses didc/the candid crate under the hood. Therefore this is an issue related to candid only. I'll ping the relevant folks.

**Comment by nomeata** [comment]
The Candid type inference algorithm in `didc` is incomplete, I think this is a known issue.

I would expect it to work if you pass the expected type using the `-t` flag to `didc --encode`.

Alternatively, you can try [`hcandid`](https://github.com/nomeata/haskell-candid), which handles this:
```
$ hcandid --encode '(
      vec {
          record { "name"; variant { TextContent = "blah" } };
          record { "ttl"; variant { Nat64Content = 1 : nat64 } };
      }
   )' | hcandid --decode --stdin
(vec {record {"name"; variant {fncrpob = "blah"}}; record {"ttl"; variant {abmwxiv = (1 : nat64)}}})
```

**Comment by chenyan-dfinity** [comment]
Yep, it's always recommended to provide the types while encoding/decoding a message. We don't have immediate plans to fix this particular issue.

Meanwhile, if you encounter a similar behavior in `dfx`, please do file a bug. Now that did file is embedded in the canister metadata, dfx should be able to fetch the did file for all canisters.

**Comment by ozwaldorf** [comment]
> Yep, it's always recommended to provide the types while encoding/decoding a message. We don't have immediate plans to fix this particular issue.
> 
> Meanwhile, if you encounter a similar behavior in `dfx`, please do file a bug. Now that did file is embedded in the canister metadata, dfx should be able to fetch the did file for all canisters.

The issue is found in the latest dfx, linked to a defined canister (but referenced by id), as where this issue was originally created :disappointed:
